### DISCUSSION TITLE
MbedTlsError(-30592) after upgrading to latest commit

### DISCUSSION THREAD

**Discussion by yanshay** [esp-rs/esp-mbedtls#57] [Q&A] [answered]
I have a program that works using `esp-mbedtls` (async)
I upgraded to the latest commit, it required adding `SHA` parameter to the `Session::new` and did the same as the example.
I started receiving `MbedTlsError(-30592)` which based on error codes I found mean: 
`0x7780    SSL - A fatal alert message was received from our peer`.

Note that I'm using TLS1.2 (that's why I'm using `esp-mbedtls`).

What's causing this?
Seems like the remote system I'm connecting to doesn't like this SHA.

Is there a way to use the new code with previous behavior? This is really critical to me.

**Comment by AnthonyGrondin** [comment]
Are you using `esp-mbedtls` as a client or as a server?

The latest changes with SHA should not affect operations, because it uses the Hardware SHA acceleration instead of the Software provided solution from `esp-mbedtls`.

**Reply by yanshay** [reply]
I'm using it as a client.
I upgraded from revision `e6e698f8613317b8a8287bb895cf88cdc63c949b`.
Maybe something else relevant changed since then? 
From my code perspective that's the only change I made.

**Reply by yanshay** [reply]
I verified, it's the last commit that breaks things. One commit before (where the SHA is not required) things work fine.
I see many things changed in that last commit.

Maybe it's the default SHA that is being used that changed?
I think the device I'm connecting to has pretty limited TLS capabilities (that's why the 1.2 is needed), so maybe it has various limitations and the client code now requires too strong encryption or something?

This is the code I'm using, it's a bit different than the example since I'm not using a certificate, maybe it has to do with that:
```

        let tls_starter = match esp_mbedtls::asynch::Session::new(
            socket,
            "",
            esp_mbedtls::Mode::Client,
            esp_mbedtls::TlsVersion::Tls1_2,
            esp_mbedtls::Certificates {
                ca_chain: None,
                ..Default::default()
            },
            tls_rx_buffer,
            tls_tx_buffer,
            &mut sha
        ) {
            Ok(tls_starter) => tls_starter.with_hardware_rsa(&mut rsa),
            Err(e) => {
                error!(boot, "Error establishing TLS Connection {:?}",e);
                Timer::after(Duration::from_millis(500)).await;
                continue; // to external loop
            }
        };

``

**Reply by AnthonyGrondin** [reply]
We don't have tests / example currently that make use of the pattern:
```rust
Certificates {
    ca_chain: None,
    ..Default::default(),
}
```

It does change the inner workings, especially for https://github.com/esp-rs/esp-mbedtls/blob/d5e29f8d89dbb96342eee91c725ce6eb91b1e02b/esp-mbedtls/src/lib.rs#L410-L419

For now, you can pick the previous commit as revision, until this get fixed. I tried locally with both async_client and sync_client and they work, so it must be something related to not providing a `ca_chain`.

For further logs, you can try and set `esp_mbedtls::set_debug()` if you are on RISCV. I believe you also need to build esp-wifi with the `sys-logs` feature enabled

**Reply by yanshay** [reply]
I'm on `esp32s3` so I believe it isn't RISCV but rather Xtensa, right?
Does that mean can't have required logs?

**Reply by AnthonyGrondin** [reply]
You can, but you'd need to patch `esp-mbedtls` due to the following line:
https://github.com/esp-rs/esp-mbedtls/blob/d5e29f8d89dbb96342eee91c725ce6eb91b1e02b/esp-mbedtls/src/lib.rs#L144-L150

This is from back when Xtensa didn't support C variadics.

**Reply by yanshay** [reply]
I opened issue #60 regarding this

**Reply by yanshay** [reply]
> You can, but you'd need to patch `esp-mbedtls` due to the following line:
> 
> https://github.com/esp-rs/esp-mbedtls/blob/d5e29f8d89dbb96342eee91c725ce6eb91b1e02b/esp-mbedtls/src/lib.rs#L144-L150
> 
> This is from back when Xtensa didn't support C variadics.

do you mean just remove the `#[cfg(not(target_arch = "xtensa"))] ` ? Now Xtensa support that?

And what integer should I set the level to?

**Reply by yanshay** [reply]
@AnthonyGrondin - I uploaded logs to the issue I opened, I wasn't able to figure out by myself anything smart from those logs except a different in size towards the end of the session.
Your assistance in resolving this will be appreciated.
Currently `esp-mbedtls` doesn't with `esp-hal` 0.22 and I wanted to try to fix it but no point doing this on an not latest version of `esp-mbedtls`.

**Comment by yanshay** [comment]
> For further logs, you can try and set esp_mbedtls::set_debug() if you are on RISCV. I believe you also need to build esp-wifi with the sys-logs feature enabled

@AnthonyGrondin I had some mixups in folders locations and seems like it didn't really take the `set_debug` properly. 
Once I fixed that I received a linker error:
`undefined reference to `mbedtls_debug_set_threshold'`

Do I need to compile myself somehow the `mbedtls` library to see logs? Is there binary available for debugging?

Any guidance/assistance would be appreciated.

**Reply by AnthonyGrondin** [reply]
Sorry about the delays,

I wrote a guide on how to compile and edit mbedtls config for each chip here: https://github.com/esp-rs/esp-mbedtls/discussions/64#discussioncomment-11753243

To fix the "undefined reference to `mbedtls_debug_set_threshold`" issue, if you're on a commit previous to 1431835 you need to rebuild mbedtls with the `MBEDTLS_DEBUG_C` flag enabled. This is done automatically past this commit

**Reply by yanshay** [reply]
I managed to get the logs for the failed session and attach them here, I don't understand much in this domain, maybe you can take a look and see if you can figure something out? As wrote on the other discussion I can't compare it with the working session.
[tls-session-bad.txt](https://github.com/user-attachments/files/18325512/tls-session-bad.txt)

**Comment by AnthonyGrondin** [comment, marked as answer]
Closing as fixed in https://github.com/esp-rs/esp-mbedtls/pull/63/commits/d6e5041ad585a930485b0352a38793255b2cf6d6

See #60 for more context
